src/bcf/mod.rs

```diff
@@ -108,6 +122,125 @@ impl Drop for Reader {
 }
 
 
+/// An indexed VCF/BCF reader.
+#[derive(Debug)]
+pub struct IndexedReader {
+    /// The synced VCF/BCF reader to use internally.
+    inner: *mut htslib::bcf_srs_t,
+    /// The header.
+    header: Rc<HeaderView>,
+
+    /// The position of the previous fetch, if any.
+    current_region: Option<(u32, u32, u32)>,
+}
+
+
+unsafe impl Send for IndexedReader {}
+
+
+impl IndexedReader {
+    /// Create a new `IndexedReader` from path.
+    ///
+    /// # Arguments
+    ///
+    /// * `path` - the path to open.
+    pub fn from_path<P: AsRef<Path>>(path: P) -> Result<Self, IndexedReaderPathError> {
+        match path.as_ref().to_str() {
+            Some(p) if path.as_ref().exists() => {
+                Ok(try!(Self::new(&ffi::CString::new(p).unwrap())))
+            },
+            _ => {
+                Err(IndexedReaderPathError::InvalidPath)
+            }
+        }
+    }
+
+    pub fn from_url(url: &Url) -> Result<Self, IndexedReaderError> {
+        Self::new(&ffi::CString::new(url.as_str()).unwrap())
+    }
+
+    /// Create a new `IndexedReader`.
+    ///
+    /// # Arguments
+    ///
+    /// * `path` - the path. Use "-" for stdin.
+    fn new(path: &ffi::CStr) -> Result<Self, IndexedReaderError> {
+        // Create reader and require existence of index file.
+        let ser_reader = unsafe { htslib::bcf_sr_init() };
+        unsafe { htslib::bcf_sr_set_opt(ser_reader, 0); } // 0: BCF_SR_REQUIRE_IDX
+        // Attach a file with the path from the arguments.
+        if unsafe { htslib::bcf_sr_add_reader(ser_reader, path.as_ptr()) } >= 0 {
+            let header = Rc::new(HeaderView::new(unsafe {
+                htslib::bcf_hdr_dup((*(*ser_reader).readers.offset(0)).header) }));
+            Ok(IndexedReader {
+                inner: ser_reader, header: header, current_region: None
+            })
+        } else {
+            Err(IndexedReaderError::InvalidPath)
+        }
+    }
+
+    pub fn fetch(&mut self, rid: u32, beg: u32, end: u32) -> Result<(), FetchError> {
+        let contig = self.header.rid2name(rid);
+        let contig = ffi::CString::new(contig).unwrap();
+        let contig = contig.as_ptr();
+        if unsafe { htslib::bcf_sr_seek(self.inner, contig, beg as i32) } != 0 {
+            Err(FetchError::Some)
+        } else {
+            self.current_region = Some((rid, beg, end));
+            Ok(())
+        }
+    }
+}
+
+
+impl Read for IndexedReader {
+    fn read(&mut self, record: &mut record::Record) -> Result<(), ReadError> {
+        match unsafe { htslib::bcf_sr_next_line(self.inner) } {
+            0 => Err(ReadError::NoMoreRecord),
+            i => {
+                assert!(i > 0, "Must not be negative");
```

Comment:
Can't this be negative in case of an error when reading the file? I guess we would like to return a `ReadError` variant in such a case, rather than having a failing assertion.

Reply:
Actually, the code counts the number of readers for which a record is found... The API is a bit weird in this area.

Reply:
And how is an error indicated?

Reply:
I don't know. Here is the usage example from the docs:

```
    Example of usage:

        bcf_srs_t *sr = bcf_sr_init();
        bcf_sr_set_opt(sr, BCF_SR_PAIR_LOGIC, BCF_SR_PAIR_BOTH_REF);
        bcf_sr_set_opt(sr, BCF_SR_REQUIRE_IDX);
        for (i=0; i<nfiles; i++)
            bcf_sr_add_reader(sr,files[i]);
        while ( bcf_sr_next_line(sr) )
        {
            for (i=0; i<nfiles; i++)
            {
                bcf1_t *line = bcf_sr_get_line(sr,i);
                ...
            }
        }
        if ( sr->errnum ) error("Error: %s\n", bcf_sr_strerror(sr->errnum));
        bcf_sr_destroy(sr);
```

Reply:
gah, it's right there...

Reply:
Ah, looks like an error leads to termination of the loop (no more record) and then you can find an error code in `sr->errnum`? So, this means here that in the `0` case you would have to check `sr->errnunm` as well.